### PR TITLE
primitives: rm redundant functions for `Transaction`

### PR DESCRIPTION
**Pull request by tcoratger**
Related https://github.com/paradigmxyz/reth/pull/11727#discussion_r1800017963

### CODE REVIEW

**Review by mattsse — CHANGES_REQUESTED** [review]
why do we need `Transaction as _`?

these prs should be done on a per function basis to make fast progress because these prs are conflict magents in case we need to change something.

I'd like to hold off on the input fn until we're sure we don't use this anywhere where this change will be expensive

**Comment by tcoratger** [comment]
> why do we need `Transaction as _`?
> 
> these prs should be done on a per function basis to make fast progress because these prs are conflict magents in case we need to change something.
> 
> I'd like to hold off on the input fn until we're sure we don't use this anywhere where this change will be expensive

@mattsse Understood, just opened a separate PR here https://github.com/paradigmxyz/reth/pull/11751, let me know if:
- You want me to do multiple PRs
- Or now that it's done keep this one and just remove what I did for the `input` function and keep it as before to prevent a bad use case?

**Comment by mattsse** [comment]
this one's fine if we don't remove the input fn

File: crates/rpc/rpc-types-compat/src/transaction/mod.rs
Comment (emhane) [reply crates/rpc/rpc-types-compat/src/transaction/mod.rs:111]: disagree, it makes sense to return a slice and let the caller clone. in many places only the slice is needed.

**Review by emhane — APPROVED** [review]

File: crates/primitives/src/transaction/mod.rs
Comment (tcoratger) [inline crates/primitives/src/transaction/mod.rs:2175]:
```suggestion
assert_eq!(tx.input().as_ref(), hex!("1b55ba3a"));
```

File: crates/rpc/rpc-types-compat/src/transaction/mod.rs
Comment (tcoratger) [inline crates/rpc/rpc-types-compat/src/transaction/mod.rs:111]:
```suggestion
    let input = tx.transaction.input().clone();
```

File: crates/rpc/rpc/src/eth/helpers/signer.rs
Comment (tcoratger) [inline crates/rpc/rpc/src/eth/helpers/signer.rs:258]:
```suggestion
        assert_eq!(Bytes::from(message.to_vec()), txn_signed.unwrap().input().0);
```

File: crates/rpc/rpc/src/eth/helpers/types.rs
Comment (tcoratger) [inline crates/rpc/rpc/src/eth/helpers/types.rs:63]:
```suggestion
                input: signed_tx.input().clone(),
```

File: crates/transaction-pool/src/traits.rs
Comment (tcoratger) [inline crates/transaction-pool/src/traits.rs:1187]:
```suggestion
        self.transaction.input().as_ref()
```

File: crates/rpc/rpc-types-compat/src/transaction/mod.rs
Comment (tcoratger) [reply crates/rpc/rpc-types-compat/src/transaction/mod.rs]: @mattsse I feel like in the few cases I changed here it was more about simplifying usage than making it heavier but you're right it's better to specify the situation for this specific `input` function and decide something in a separate PR. So I reverted the changes made to this function so we can merge the PR and then see what we do for `input`.

**Comment by tcoratger** [comment]
> this one's fine if we don't remove the input fn

@mattsse done, ready to review/merge

File: crates/rpc/rpc-types-compat/src/transaction/mod.rs
Comment (mattsse) [reply crates/rpc/rpc-types-compat/src/transaction/mod.rs]: this matters when we no longer need the slice but need to get `Bytes`, then this suddenly becomes expensive for no reason.

I agree that we can keep input -> slice but should have another function that returns `Bytes` for this usecase

**Comment by mattsse** [comment]
pedantic clippy, otherwise lgtm

**Comment by tcoratger** [comment]
> pedantic clippy, otherwise lgtm

@mattsse should be fixed

File: crates/rpc/rpc-types-compat/src/transaction/mod.rs
Comment (tcoratger) [reply crates/rpc/rpc-types-compat/src/transaction/mod.rs]: Yeah was thinking about the same dual approach, will open a follow up PR directly after this one is merged

**Review by mattsse — APPROVED** [review]

**Comment by mattsse** [comment]
@tcoratger this drifted slightly and needs a small rebase

**Comment by tcoratger** [comment]
> @tcoratger this drifted slightly and needs a small rebase

@mattsse should be fixed, ready to merge :)

**Pull request merged**
